keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1108 @@
+not be tied to Kubernetes releases. Below, we include a notional schedule for development.
+
+- Kubernetes 1.10 - Alpha CRD and controller implementation
+- Kuberentes 1.11 - Beta CRD and controller implementation
+
+## Drawbacks
+
+### Increased API Server Storage Requirements
+As the Application object will be stored on the API Server, if the users create many Applications, the API Server 
+(i.e. etcd) will have an increased storage requirement. Users already create many ConfigMaps and Secrets. It is worth
+pointing out that if Application metadata is stored on every object using label and annotation conventions, the 
+storage burden on the API Server will likely be greater than if we aggregate the information in a single object and 
+store it once. Also, users are already using CRDs for a similar function, and, for those users, the API Server already 
+tolerates the additional storage burden.
+
+### Users Have to Create an Application Object
+While users do have to create an additional object to make use of the functionality described in this proposal, we 
+endeavor to keep the burden small, and the resource simple. The example manifests above demonstrate a minimal 
+Application object. We also expect that tools like Helm and Konflate will be able to auto-generate the App resource 
+from existing manifests.
+
+## Alternatives 
+This section contains various design alternatives that were considered during the generation of this KEP.
+
+### Use Labeling Conventions
+Developers should use the accepted Kubernetes 
+[labeling and annotation conventions](https://docs.google.com/document/d/1EVy0wRJRm5nogkHl38fNKbFrhERmSL_CLNE4cxcsc_M). 
+The Application API is complementary to, and an extension of, this approach. Users that use the Application resource in 
+addition to the accepted labeling convention will be able to make use of the functionality described in this 
+document.
+
+### Application Type Aggregation
+An Application object must be created for each instance of application and not each type of application. It 
+might be useful to provide an object that aggregates Applications by type (e.g. all instances of WordPress that are 
+installed in the system), but this can also be accomplished by querying for all Applications and filtering or 
+aggregating based on the declared `.Spec.Type` field. In the future, if the Application Kind is widely adopted, it would 
+be possible to extend the system with another Kind that provides this aggregation via the API object model. The 
+challenges here would be multi-tenancy and namespace isolation. If a higher level object aggregates all Applications 
+of the same Type, it would have to do so across namespaces. This is not compatible with the existing namespace model 
+for Kubernetes unless we assume that the higher level object is only accessible to cluster administrators with 
+permissions to access all namespaces in the cluster.
+
+## Future Work
```

Comment:
It would be nice to aggregate events that come from the same app.  Not sure if app controller is the right thing to do that.